### DISCUSSION TITLE
Many handlers in hoop

### DISCUSSION THREAD

**Discussion by josejachuf** [salvo-rs/salvo#210] [Q&A] [answered]
Hi, 

How to inject several middleware to the same route?

**Comment by chrislearn** [comment, marked as answer]
```rust
router.hoop(middleware1).hoop(middleware2)
```

**Reply by josejachuf** [reply]
Thank! Very simple!

Another question about it.

I have a middleware with the pool of connections almost in the trunk of the route tree:

`.hoop(affix::inject(db_pool))`

then in function I get the pool with:

`let pg_pool = depot.obtain::<PgPool>().unwrap();`

This work fine.

Now if I add another middleware (session) before Pool's

```
let session_handler = SessionHandler::builder(
        MS::new(),
        b"secretsecretsecretsecretsecretsecretsecretsecretsecretsecretsecretsecretsecret")
            .build()
            .unwrap();
```

> MS is MemoryStore, I rename because cause confict with cache MemoryStore
`
Router::new().hoop(session_handler)`

```
#[handler]
pub async fn my_function(req: &mut Request, depot: &mut Depot, res: &mut Response) {
    let id = req.param::<i64>("id").unwrap_or(0);
    let pg_pool = depot.obtain::<PgPool>().unwrap();
    // Here
```

How to recover **handler_session** at my_function?

I try this:
    
    if let Some(session) = depot.session_mut() {
         if let Some(username) = session.get::<String>("username") {
             println!("Hello, {}. ", username);
         }
    }


but I get this error:



 let pg_pool = depot.obtain::<PgPool>().unwrap();
   |                   ------------------------ immutable borrow occurs here
28 |     if let Some(session) = depot.session_mut() {
   |                            ^^^^^^^^^^^^^^^^^^^ mutable borrow occurs here
...
36 |         .fetch_one(pg_pool)
   |                    ------- immutable borrow later used here

**Comment by josejachuf** [comment]
changing **depot.session_mut()** **for depot.session()** is solved

```
#[handler]
pub async fn my_function(req: &mut Request, depot: &mut Depot, res: &mut Response) {
    let id = req.param::<i64>("id").unwrap_or(0);
    let pg_pool = depot.obtain::<PgPool>().unwrap();
    if let Some(session) = depot.session() {
        if let Some(username) = session.get::<String>("username") {
            println!("Hello, {}. ", username);
        }
    }

}
```
